### PR TITLE
feat: extends manifest to allow for preview features

### PR DESCRIPTION
**Pull request by tdejager**
### Description

This PR adds the ability to allow the use of preview features, we want to use this to enable certain features that should be enabled on the manifest level. The first one being, the workspace and build functionality. The idea behind these features is that we can still break and test out new workflows while being in preview, but we do like to get some hands-on testing for these features.

Say for example we allow for a new solver (don't worry we are not really doing this :)), you could then specify:

```toml
[project]
# ... more metadata
preview = ["new_solver"]
```

Alternatively, to allow for all preview features one can use `preview = true`. 

### Notes

I've modified the schema and the documentation for this change. No features have been specified as of yet, but it should be relatively straightforward by adding an enum value to the `KnownPreviewFeatures` enum.

### Testing

I think most of the testing for this is covered in CI for now. Best to look at the API in project and see if this is something that looks good for now.

### CODE REVIEW

File: schema/model.py [hunk 1]
```diff
@@ -139,6 +143,9 @@ class Project(StrictBaseModel):
     pypi_options: PyPIOptions | None = Field(
         None, alias="pypi-options", description="Options related to PyPI indexes for this project"
     )
+    preview: list[KnownPreviewFeature | str] | bool | None = Field(
```
Comment (baszalmstra) [inline schema/model.py:146]: Instead of a `class` I think it makes more sense to use `Literal["new_solve"]`

File: src/project/mod.rs [hunk 2]
```diff
@@ -955,6 +956,21 @@ impl Project {
...
+
+    /// Returns true if all preview features are enabled
+    pub fn all_preview_features_enabled(&self) -> bool {
+        self.manifest
+            .preview()
+            .map(|preview| preview.all_enabled())
+            .unwrap_or(false)
+    }
+
+    pub fn is_preview_feature_enabled(&self, feature: &KnownPreviewFeature) -> bool {
```
Comment (baszalmstra) [inline src/project/mod.rs:968]: I assume `KnownPreviewFeature` is a `Copy` type so there is no need to pass it by reference!

File: schema/model.py (hunk 1 above)
Comment (tdejager) [reply schema/model.py:146]: But I can't make an empty version for the mean time right?

File: crates/pixi_manifest/src/validation.rs [hunk 3]
```diff
@@ -132,6 +132,22 @@ impl ParsedManifest {
             }
         }
 
+        // Warn on any unknown preview features
+        if let Some(preview) = self.project.preview.as_ref() {
+            let preview = preview.unknown_preview_features();
```
[the remark this answers is not in the document]
Comment (ruben-arts) [reply crates/pixi_manifest/src/validation.rs:137]: Tip, you could use serde_ignored for this logic. 
Check the `pixi_config` crate : https://github.com/prefix-dev/pixi/blob/14108646d73a198b5c96a428f998497c87442f8a/crates/pixi_config/src/lib.rs#L557-L568

**Review by ruben-arts — COMMENTED** [review]
Looks good, a bit hard to test as there is no preview feature yet. But works as expected so far.

I do think it's good to use `serde_ignored` for the warning to streamline that workflow over pixi's parsing.

File: crates/pixi_manifest/src/validation.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tdejager) [reply crates/pixi_manifest/src/validation.rs:137]: Wow sick let me check that out :)

[the remark this answers is not in the document]
Comment (tdejager) [reply crates/pixi_manifest/src/validation.rs:137]: That did not seem to work, alas it only works on keys :)

**Comment by tdejager** [comment]
Also checked out if I could do with just an `#[serde(untagged)]` in the enum list, however, this messes up the error messages again, and flattening `KnownPreviewFeature` and `PreviewFeature` had a bit of bad DX when adding and removing from the enum which we would probably do quite a bit.

Now you can just add/remove to the enum and everything just works. So would rather keep like this.

**Pull request merged**